crates/dbsp/src/algebra/decimal.rs

```diff
@@ -0,0 +1,607 @@
+        let max_precision = Precision::max();
+
+        let max_scale = Scale::try_from(scale)?;
+
+        inner.rescale(max_scale.inner());
+
+        Ok(Self {
+            inner,
+            max_precision,
+            max_scale,
+        })
+    }
+
+    pub fn rescale(mut self, val: u32) -> Result<Self, String> {
+        self.max_scale = Scale::try_from(val)?;
+        self.max_scale.apply(&mut self.inner);
+
+        Ok(self)
+    }
+
+    pub fn precision(mut self, val: u32) -> Result<Self, String> {
+        self.max_precision = Precision::try_from(val)?;
+
+        Ok(self)
+    }
+
+    fn apply_scaling_rules(&mut self) {
+        self.max_scale.apply(&mut self.inner)
+    }
+
+    pub fn trunc(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.trunc();
+        out.apply_scaling_rules();
+        out
+    }
+
+    pub fn rescale_with_prec(&mut self, precision: u32, scale: i32) {
+        let scale = scale as u32;
+        self.inner.rescale(scale);
+
+        self.max_scale = Scale::try_from(self.inner.scale()).unwrap();
+        self.max_precision = Precision::try_from(precision).unwrap();
+    }
+
+    pub fn ln(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.ln();
+        out.apply_scaling_rules();
+        out
+    }
+
+    pub fn log10(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.log10();
+        out.apply_scaling_rules();
+        out
+    }
+
+    pub fn round_dp(&self, val: u32) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.round_dp(val);
+        out.apply_scaling_rules();
+        out
+    }
+
+    pub fn trunc_with_scale(&self, scale: u32) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.trunc_with_scale(scale);
+        out.apply_scaling_rules();
+        out
+    }
+
+    pub fn sqrt(&self) -> Option<Self> {
+        let mut out = *self;
+        out.inner = out.inner.sqrt()?;
+        out.inner.normalize_assign();
+        out.max_scale = Scale::try_from(out.inner.scale()).unwrap_or(Scale::max());
+        out.apply_scaling_rules();
+        Some(out)
+    }
+
+    pub fn powd(&self, exp: SQLDecimal) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.powd(exp.inner);
+        out.apply_scaling_rules();
+        out
+    }
+
+    pub fn floor(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.floor();
+        out.max_scale = Scale::min();
+        out.apply_scaling_rules();
+        out
+    }
+
+    pub fn ceil(&self) -> Self {
+        let mut out = *self;
+        out.inner = out.inner.ceil();
+        out.max_scale = Scale::min();
+        out.apply_scaling_rules();
+        out
+    }
+}
+
+impl One for SQLDecimal {
+    fn one() -> Self {
+        Self::try_from(Decimal::ONE).unwrap()
+    }
+}
+
+impl Rem for SQLDecimal {
+    type Output = Self;
+
+    fn rem(self, rhs: Self) -> Self::Output {
```

Comment:
what is rem?

Reply:
Is this defined for non-integers?

Reply:
why no scaling rules here?

Reply:
Rem is the trait for the modulo operator. It is not implemented for floats in the standard library but the `rust_decimal::Decimal` type implements it. 

Scaling rules weren't applied here because they produce outputs with scale and precision different to the input.